Публикация постов по расписанию

SAM19991 opened the pull request:
Создан класс исключения ScheduledPostPublicationException для обработки ошибок при публикации запланированных постов

Доработан PostServiceImpl для поддержки отложенной публикации:

Добавлен метод publishScheduledPosts() для поиска и публикации запланированных постов Реализовано пакетное выполнение публикации через createBatches() Добавлена асинхронная обработка с использованием CompletableFuture и отдельного пула потоков

Добавлено планирование задач через аннотацию @Scheduled с настройкой на выполнение каждую минуту (class ScheduledPostPublisher)

В PostDto добавлено поле scheduledAt типа LocalDateTime

SAM19991 committed:
Создан класс исключения ScheduledPostPublicationException для обработки ошибок при публикации запланированных постов

Доработан PostServiceImpl для поддержки отложенной публикации:

Добавлен метод publishScheduledPosts() для поиска и публикации запланированных постов
Реализовано пакетное выполнение публикации через createBatches()
Добавлена асинхронная обработка с использованием CompletableFuture и отдельного пула потоков

Добавлено планирование задач через аннотацию @Scheduled с настройкой на выполнение каждую минуту (class ScheduledPostPublisher)

В PostDto добавлено поле scheduledAt типа LocalDateTime

kashkademan commented on src/main/java/faang/school/postservice/dto/post/PostDto.java:
```diff
@@ -10,10 +11,15 @@ public record PostDto(
         Long projectId,
-        List<String> resourceKeys
+        List<String> resourceKeys,
+        LocalDateTime scheduledAt
```
Это норм, что в DTO объектом? спринг завернет это в json?

SAM19991 replied:
В задании так было написано (В PostDto нужно добавить новое поле: LocalDateTime scheduledAt. Оно будет показывать, на какие дату и время запланирован этот пост.)

kashkademan commented on src/main/java/faang/school/postservice/service/PostServiceImpl.java:
```diff
@@ -5,30 +5,38 @@
     private final ProjectServiceClient projectServiceClient;
     private final ResourceService resourceService;
+    private final ExecutorService scheduledPostExecutorService;
+    private static final int BATCH_SIZE = 1000;
```
вроде как рекомендация была не использовать статические переменные... но в целом так не к тому, чтобы править.

SAM19991 replied:
Я видимо пропустил эту рекомендацию, а как мне ее заменить тогда?

mysosnovskaya replied:
Брать из application.yaml через аннотацию Value

kashkademan commented on src/main/java/faang/school/postservice/service/PostServiceImpl.java:
```diff
@@ -177,4 +185,45 @@ private void validateAuthor(Long authorId, Long projectId) {
+
+    private List<List<Post>> createBatches(List<Post> posts) {
+        return IntStream.range(0, (posts.size() + BATCH_SIZE - 1) / BATCH_SIZE)
+                .mapToObj(i -> posts.subList(i * BATCH_SIZE, Math.min((i + 1) * BATCH_SIZE, posts.size()))).toList();
```
сложная штука... делишь на батчи пропорционально чему-то?

SAM19991 replied:
Вроде сейчас понятнее

SAM19991 committed:
Улучшил метод пакетной обработки постов для большей читаемости кода.

SAM19991 committed:
Улучшил метод пакетной обработки постов для большей читаемости кода.

SAM19991 committed:
Улучшил метод пакетной обработки постов для большей читаемости кода.

SAM19991 committed:
Улучшил метод пакетной обработки постов для большей читаемости кода.

SAM19991 committed:
Улучшил метод пакетной обработки постов для большей читаемости кода.

mysosnovskaya commented on src/main/java/faang/school/postservice/config/ThreadPoolConfig.java:
```diff
@@ -0,0 +1,19 @@
+    @Bean
+    public ExecutorService scheduledPostExecutorService() {
+      log.info("Creating thread pool with 10 threads for scheduled post publishing");
+      return Executors.newFixedThreadPool(10);
```
10 - в application.yaml

mysosnovskaya commented on src/main/java/faang/school/postservice/scheduler/ScheduledPostPublisher.java:
```diff
@@ -0,0 +1,26 @@
+
+    private final PostService postService;
+
+    @Scheduled(cron = "0 * * * * *")
```
cron - в application.yaml

mysosnovskaya commented on src/main/java/faang/school/postservice/service/PostServiceImpl.java:
```diff
@@ -177,4 +185,51 @@ private void validateAuthor(Long authorId, Long projectId) {
+        log.info("Finished scheduled post publishing job");
+    }
+
+    private List<List<Post>> createBatches(List<Post> posts) {
```
Есть уже готовые решения для этого. Используй ListUtils.partition от apache (нужная зависимость уже должна быть в проекте)

mysosnovskaya commented on src/main/java/faang/school/postservice/service/PostServiceImpl.java:
```diff
@@ -177,4 +185,51 @@ private void validateAuthor(Long authorId, Long projectId) {
+            log.debug("Successfully published {} posts", batch.size());
+        } catch (Exception e) {
+            log.error("Error while publishing scheduled posts", e);
+            throw new ScheduledPostPublicationException("Failed to publish scheduled posts", e.getCause());
```
Кто будет перехватывать это исключение? И будет ли оно логироваться второй раз (первый раз ты логируешь самостоятельно на предыдущей строке)?

SAM19991 committed:
Конфигурация:
- Вынесены хардкод значения в : размер батча (1000), размер пула потоков (10), cron-выражение `application.yaml`
- Добавлены `@Value` аннотации для чтения конфигурации

Упрощение кода:
- Заменен сложный метод `createBatches()` на `ListUtils.partition()`
- Удалена константа `BATCH_SIZE`

Исправления:
- Улучшено логирование ошибок с указанием размера батча
- Исправлена передача исключения (убран `.getCause()`)

**Добавлены импорты:**
- `@Value`, `ListUtils`

SAM19991 committed:
Merge branch 'triton-master-stream10' into develop/BJS2-77875

# Conflicts:
#	src/main/java/faang/school/postservice/service/PostServiceImpl.java
#	src/main/resources/application.yaml

SAM19991 committed:
Конфигурация:
- Вынесены хардкод значения в : размер батча (1000), размер пула потоков (10), cron-выражение `application.yaml`
- Добавлены `@Value` аннотации для чтения конфигурации

Упрощение кода:
- Заменен сложный метод `createBatches()` на `ListUtils.partition()`
- Удалена константа `BATCH_SIZE`

Исправления:
- Улучшено логирование ошибок с указанием размера батча
- Исправлена передача исключения (убран `.getCause()`)

mysosnovskaya commented on src/main/java/faang/school/postservice/service/PostServiceImpl.java:
```diff
@@ -209,4 +217,45 @@ private void validateAuthor(Long authorId, Long projectId) {
+        log.info("Finished scheduled post publishing job");
+    }
+
+    private List<List<Post>> createBatches(List<Post> posts) {
```
Как будто бы отдельный метод теперь потерял свой смысл, т.к. он из одной строки всего состоит)

SAM19991 replied:
Ты абсолютно права

mysosnovskaya commented on src/main/java/faang/school/postservice/service/PostServiceImpl.java:
```diff
@@ -209,4 +217,45 @@ private void validateAuthor(Long authorId, Long projectId) {
+            postRepository.saveAll(batch);
+            log.debug("Successfully published {} posts", batch.size());
+        } catch (Exception e) {
+            log.error("Error while publishing scheduled posts in batch of size {}", batch.size(), e);
+            throw new ScheduledPostPublicationException("Failed to publish scheduled posts", e);
```
Продублирую свои вопросы, т.к. не получила ответы:
1. Не будет ли двойного логирования?
2. Точно ли нужно выбрасывать это исключение? Для чего? На что это повлияет?

SAM19991 replied:
Убрал все лишнее т.к. было бы двойное логирование и один неудачный батч сломал бы публикацию всех остальных

SAM19991 committed:
- Удалил ненужный метод createBatches

- Встроили ListUtils.partition() напрямую в основной метод

- Ведем счетчики успешных/неудачных батчей
- Логирование итоговой статистики  выполнения задачи
- Один неудачный батч больше не останавливает публикацию остальных постов
